### PR TITLE
Adds generic Query parameter to next.js's router

### PR DESCRIPTION
**Pull request by kaw2k**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [ ] ~Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>~
- [ ] Increase the version number in the header if appropriate.
^ *Not sure if this is located* 
- [ ] ~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.~

I was looking for a simple solution to get static typings for the `query` object passed down from the router. With next using the query object instead of route params, it seems to make sense to give some option to type it, if for nothing else than documentation :D

Proposed useage would be like this:

```tsx
import { withRouter, WithRouterProps } from 'next/router'

interface Query {
  blogId?: string
}

// The new generic type is optional, defaulting to the existing behavior
interface Props extends WithRouterProps<Query> {}

const Blog: React.SFC<Props> = ({ router }) => {
  // Here router.query autocompletes to blogId
  return <div>Blog {router.query && router.query.blogId}</div>
}

export default withRouter(Blog)
```

Thoughts? I didn't find a clean way to type the query object and thought this may be useful for others too. 

One small snag, I wasn't able to add the generic to `withRouter` without the linter yelling at me for single use generics. If you like the idea, I will play more to get `withRouter` typed similarly.

Perhaps another place to add this would be the `NextContext` object as well as the query is passed to `getInitialProps`.

### CODE REVIEW

File: types/next/router.d.ts [hunk 1]
```diff
@@ -78,14 +80,14 @@ export interface RouterProps {
...
     ready(cb: RouterCallback): void;
 }
 
-export interface WithRouterProps {
-    router: SingletonRouter;
+export interface WithRouterProps<Q = DefaultQuery> {
+    router: SingletonRouter<Q>;
 }
 
 export function withRouter<T extends {}>(
```
Comment (resir014) [inline types/next/router.d.ts]: To fix the single-use generic issue on `dtslint` you may be able to replace `withRouter` with the following:

```ts
export const withRouter: WithRouterFn;
```

and including `WithRouterFn` as a separate, exportable type:

```ts
export type WithRouterFn = <T extends {}, Q = DefaultQuery>(Component: React.ComponentType<T & WithRouterProps<Q>>): React.ComponentType<T>;
```

**Review by resir014 — COMMENTED** [review]
Thanks a lot for this, you rock!

I've left a note re: the `dtslint` issue, but overall LGTM. 👍

File: types/next/router.d.ts (hunk 1 above)
Comment (kaw2k) [reply types/next/router.d.ts]: Thank you for the idea! I tried it out and the issue persisted. I can understand the error, however, this is a weird use case where we have it present for type inference in (maybe) rare use cases. I opted for overriding the linter and would be more than happy to go back and poke more.

**Comment by kaw2k** [comment]
I added the same generic for the `NextContext` as it would be equally as useful. I am playing around with this more, want to take some time to make sure nothing is missed and no sneaky bugs sneak through. After everything is finished and reviewed I will squash all the commits into one for the merge.

**Comment by DanielRosenwasser** [comment]
:bell: @Dru89 @brikou @jthegedus @resir014 @scottdj92 @joaovieira any thoughts on this one?

**Review by joaovieira — APPROVED** [review]
This looks alright, and I've suffered the same before. Even though I'm wondering about the initial `DefaultQuery` type as Node's `querystring.parse` method only returns `string | string [] | undefined`. The same is used in the [`NextContext` query](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/next/index.d.ts#L21):

https://github.com/zeit/next.js/blob/master/lib/router/router.js#L161
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v8/index.d.ts#L899

**Review by resir014 — APPROVED** [review]

**Review by resir014 — APPROVED** [review]

**Review by jthegedus — APPROVED** [review]

**Pull request merged**